x-pack/plugins/actions/server/cleanup_failed_executions/find_and_cleanup_tasks.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Logger, CoreStart } from 'kibana/server';
+import { ActionsConfig } from '../config';
+import { ActionsPluginsStart } from '../plugin';
+import { ActionTypeRegistryContract } from '../types';
+import { cleanupTasks, CleanupTasksResult } from './cleanup_tasks';
+import { TaskInstance } from '../../../task_manager/server';
+import { nodeBuilder } from '../../../../../src/plugins/data/common';
+
+export interface FindAndCleanupTasksOpts {
+  logger: Logger;
+  actionTypeRegistry: ActionTypeRegistryContract;
+  coreStartServices: Promise<[CoreStart, ActionsPluginsStart, unknown]>;
+  config: ActionsConfig['cleanupFailedExecutionsTask'];
+  kibanaIndex: string;
+  taskManagerIndex: string;
+}
+
+export interface FindAndCleanupTasksResult extends CleanupTasksResult {
+  remaining: number;
+}
+
+export async function findAndCleanupTasks({
+  logger,
+  actionTypeRegistry,
+  coreStartServices,
+  config,
+  kibanaIndex,
+  taskManagerIndex,
+}: FindAndCleanupTasksOpts): Promise<FindAndCleanupTasksResult> {
+  logger.debug('Starting cleanup of failed executions');
+  const [{ savedObjects, elasticsearch }, { spaces }] = await coreStartServices;
+  const esClient = elasticsearch.client.asInternalUser;
+  const savedObjectsClient = savedObjects.createInternalRepository(['task']);
+  const savedObjectsSerializer = savedObjects.createSerializer();
+
+  const result = await savedObjectsClient.find<TaskInstance>({
+    type: 'task',
+    filter: nodeBuilder.and([
+      nodeBuilder.is('task.attributes.status', 'failed'),
+      nodeBuilder.or(
+        actionTypeRegistry
+          .list()
+          .map((actionType) =>
+            nodeBuilder.is('task.attributes.taskType', `actions:${actionType.id}`)
```

Comment:
I guess this means we won't clean up anything but failed actions?  Seems ok for now, but I suspect we'll end up with some other task types in the future which also leave tombstones.

Reply:
Yeah, agreed. I felt safe doing so for actions to begin with since they are traced with the event log while other ad-hoc tasks aren't (yet).